protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractMailboxProcessor.java

```diff
@@ -584,10 +585,11 @@ protected void respondVanished(MailboxSession session, MessageManager mailbox, L
             }
             searchQuery.andCriteria(SearchQuery.uid(nRanges));
             searchQuery.andCriteria(SearchQuery.modSeqGreaterThan(changedSince));
-            Iterator<MessageUid> uids = mailbox.search(searchQuery, session);
+            CloseableIterator<MessageUid> uids = mailbox.search(searchQuery, session);
             while (uids.hasNext()) {
                 vanishedUids.remove(uids.next());
             }
+            uids.closeSilently();
```

Comment:
look like you need to use `closeSilently()` every time with this `CloseableIterator`, how about implement a new `SilentlyCloseableIterator` and override `close()` method to do what the `closeSilently()` does?

Then you can use try with resource

Reply:
I don't like the idea of letting such a low level detail leak in our abstraction and enforce behaviours like that.

Does it makes sense that the `MessageManager` forces you, using the type system, to sneaky throw? It's none of its business.

However you are right to point out the fact that try-with-resources should be used - it avoids exceptions bypassing the close.

Reply:
I found a subtile way of implementing your proposal while not enforcing the caller behaviour within the return type.